src/mlpack/methods/mean_shift/mean_shift_impl.hpp

```diff
@@ -0,0 +1,129 @@
+/**
+ * @file mean_shift_impl.hpp
+ * @author Shangtong Zhang
+ *
+ * Mean Shift clustering
+ */
+
+#include "mean_shift.hpp"
+#include <mlpack/core/kernels/gaussian_kernel.hpp>
+
+namespace mlpack {
+namespace meanshift {
+  
+/**
+  * Construct the Mean Shift object.
+  */
+template<typename KernelType,
+         typename MatType>
+MeanShift<
+  KernelType,
+  MatType>::
+MeanShift(const double duplicateThresh,
+          const size_t maxIterations,
+          const double stopThresh,
+          const KernelType kernel) :
+    duplicateThresh(duplicateThresh),
+    maxIterations(maxIterations),
+    stopThresh(stopThresh),
+    kernel(kernel)
+{
+  // Nothing to do.
+}
+
+/**
+  * Perform Mean Shift clustering on the data, returning a list of cluster
+  * assignments and centroids.
+  */
+template<typename KernelType,
+         typename MatType>
+inline void MeanShift<
+    KernelType,
+    MatType>::
+Cluster(const MatType& data,
+        arma::Col<size_t>& assignments,
+        arma::mat& centroids) {
+  
+  // all centroids before remove duplicate ones.
+  arma::mat allCentroids(data.n_rows, data.n_cols);
+  assignments.set_size(data.n_cols);
+  
+  // for each point in dataset, perform mean shift algorithm.
+  for (size_t i = 0; i < data.n_cols; ++i) {
+    
+    size_t completedIterations = 0;
+    
+    //initial centroid is the point itself.
+    allCentroids.col(i) = data.col(i);
+    
+    while (true) {
+      
+      // new centroid
+      arma::Col<double> newCentroid = arma::zeros(data.n_rows, 1);
+      
+      double sumWeight = 0;
+      for (size_t j = 0; j < data.n_cols; ++j) {
+        
+        // calc weight for each point
+        double weight = kernel.Evaluate(allCentroids.col(i), data.col(j));
```

Comment:
This is only correct for the Gaussian kernel, because d/dx(exp(-x/h)) = -1/h exp(-x/h).  For other kernels the expression g(x) might be substantially different (my notation comes from http://homepages.inf.ed.ac.uk/rbf/CVonline/LOCAL_COPIES/TUZEL1/MeanShift.pdf ).  I would suggest writing a standalone function that calculates the weight for a point, and then using a template specialization for the `GaussianKernel` and other kernels where you can just use the kernel evaluation.  In cases where the kernel is not radially symmetric, the case may be even more different (I have not derived that case myself).  Otherwise, you will probably have to require that each kernel has a `Gradient()` function and then you can call that when one of the template specializations is not being used.